평균 제곱근 오차(Root Mean Squared Error, RMSE)  
임의의 선을 그리고 난 후 이 선이 얼마나 잘 그려졌는지 평가해서 조금씩 수정해나간다.  
이를 위해서 주어진 선의 오차를 평가하는 알고리즘이 필요한데 그 중에서 가장 많이 사용되는 방법이 평균 제곱근 오차이다.

In [1]:
import warnings
warnings.filterwarnings(action='ignore')
%config Completer.use_jedi = False
import numpy as np
import pandas as pd

In [2]:
# x, y의 데이터 값이 1건씩 저장된 리스트 => [공부시간(x), 시험점수(y)]
data = [[2, 81], [4, 93], [6, 91], [8, 97]]
# 기울기 a와 y절편 b를 임의로 정한다. => [기울기, y절편]
ab = [2.3, 79] # 2.3, 79

for i in data:
    print('공부시간: {}, 시험성적: {}'.format(i[0], i[1]))

공부시간: 2, 시험성적: 81
공부시간: 4, 시험성적: 93
공부시간: 6, 시험성적: 91
공부시간: 8, 시험성적: 97


In [3]:
# x, y의 데이터 값이 1건씩 저장된 리스트에서 공부시간(x), 시험성적(y)을 뽑아내서 각각 별도의 리스트에 저장한다.
x = [i[0] for i in data]
y = [i[1] for i in data]
for i in range(len(data)):
    print('공부시간: {}, 시험성적: {}'.format(x[i], y[i]))

공부시간: 2, 시험성적: 81
공부시간: 4, 시험성적: 93
공부시간: 6, 시험성적: 91
공부시간: 8, 시험성적: 97


In [4]:
# Y = a * x + b에 a(기울기), b(y절편) 값을 대입해서 결과(예측성적)를 계산하는 함수
def predict(x):
    return ab[0] * x + ab[1]

In [5]:
# 임의로 정한 기울기와 y절편으로 예측성적을 계산한다.
# 임의로 정한 기울기와 y절편을 이용해 계산된 예측 성적이 저장될 빈 리스트를 만든다.
predict_result = []
# 모든 공부시간(x) 값을 한 번씩 대입해서 예측성적 리스트(predict_result)를 완성한다.
for i in range(len(x)):
    # 공부시간에 따른 예측성적을 계산하는 함수 predict()를 실행해서 얻어진 예측성적을 predict_result에 추가한다.
    predict_result.append(predict(x[i]))
    print('공부시간: {}, 시험성적: {}, 예측성적: {}'.format(x[i], y[i], predict_result[i]))

공부시간: 2, 시험성적: 81, 예측성적: 83.6
공부시간: 4, 시험성적: 93, 예측성적: 88.2
공부시간: 6, 시험성적: 91, 예측성적: 92.8
공부시간: 8, 시험성적: 97, 예측성적: 97.4


In [6]:
# 평균 제곱근 오차를 계산하는 함수
# rmse(예측성적, 실제성적) => 인수로 넘겨받는 데이터는 넘파이 배열 타입의 데이터이다.
def rmse(predict_result, y):
    # mean(): numpy에서 평균을 계산한다.
    # sqrt(): numpy에서 제곱근을 계산하는 함수
    return np.sqrt(((predict_result - y) ** 2).mean())

In [7]:
# 평균 제곱근 오차를 계산할 데이터(list)를 넘겨받아 numpy 배열로 만들어 평균 제곱근 오차를 계산하는 함수 rmse()를
# 호출하는 함수
def rmse_val(predict_result, y):
    # array(): 파이썬의 list를 numpy 배열로 변환한다.
    # numpy 배열로 변환된 예측성적과 실제성적으로 평균 제곱근 오차를 계산하는 함수를 실행해서 결과를 리턴시킨다.
    return rmse(np.array(predict_result), np.array(y))

In [8]:
# 최종 평균 제곱근 오차를 출력한다.
rmse_result = rmse_val(predict_result, y)
print('RMSE: {}'.format(rmse_result))

RMSE: 2.880972058177584


기울기 2.8, y절편 73일 경우 RMSE는 4.6690470119715가 계산되는 것을 알 수 있다.  
기울기 3, y절편 76일 경우 RMSE는 3.3166247903554가 계산되는 것을 알 수 있다.  
이 오차를 줄이면서 새로운 선을 긋는 일이다.  
이를 위해서는 기울기와 y절편을 적절히 조정하면서 오차의 변화를 살펴보고 오차가 최소가 되는 기울기와 y절편을 찾아야 한다.  
04_Method_of_Least_Squares_최소제곱법 예제에서 계산했던 기울기와 y절편을 사용하면 2.880972058177584의 RMSE가 발생된다.